flink-connectors/flink-connector-elasticsearch7/src/main/java/org/apache/flink/connector/elasticsearch/source/reader/Elasticsearch7SplitReader.java

```diff
@@ -0,0 +1,238 @@
+import org.apache.flink.connector.elasticsearch.source.Elasticsearch7SourceConfiguration;
+import org.apache.flink.connector.elasticsearch.source.split.Elasticsearch7Split;
+import org.apache.flink.util.Collector;
+
+import org.elasticsearch.search.SearchHit;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.Nullable;
+
+import java.io.IOException;
+import java.util.ArrayDeque;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Queue;
+import java.util.Set;
+
+/**
+ * A {@link SplitReader} implementation that reads {@link Elasticsearch7Record}s from {@link
+ * Elasticsearch7Split}s.
+ *
+ * @param <T> the type of the record to be emitted from the Source.
+ */
+public class Elasticsearch7SplitReader<T>
+        implements SplitReader<Elasticsearch7Record<T>, Elasticsearch7Split> {
+    private static final Logger LOG = LoggerFactory.getLogger(Elasticsearch7SplitReader.class);
+    private final Elasticsearch7SourceConfiguration sourceConfiguration;
+    private final NetworkClientConfig networkClientConfig;
+    private final Queue<Elasticsearch7Split> splits;
+
+    @Nullable private Elasticsearch7SearchHitReader currentReader;
+    @Nullable private String currentSplitId;
+
+    private final Elasticsearch7SearchHitDeserializationSchema<T> deserializationSchema;
+    private final SimpleCollector<T> collector;
+
+    public Elasticsearch7SplitReader(
+            Elasticsearch7SourceConfiguration sourceConfiguration,
+            NetworkClientConfig networkClientConfig,
+            Elasticsearch7SearchHitDeserializationSchema<T> deserializationSchema) {
+        this.sourceConfiguration = sourceConfiguration;
+        this.networkClientConfig = networkClientConfig;
+        this.deserializationSchema = deserializationSchema;
+        this.splits = new ArrayDeque<>();
+        this.collector = new SimpleCollector<>();
+    }
+
+    private RecordsWithSplitIds<Elasticsearch7Record<T>> createRecordsFromSearchResults(
+            Collection<SearchHit> searchHits) throws IOException {
+        Collection<Elasticsearch7Record<T>> recordsForSplit = new ArrayList<>();
+        for (SearchHit hit : searchHits) {
+            try {
+                deserializationSchema.deserialize(hit, collector);
+                collector
+                        .getRecords()
+                        .forEach(r -> recordsForSplit.add(new Elasticsearch7Record<>(r)));
+            } catch (Exception e) {
+                throw new IOException("Failed to deserialize consumer record due to", e);
+            } finally {
+                collector.reset();
+            }
+        }
+
+        return ElasticsearchSplitRecords.forRecords(currentSplitId, recordsForSplit);
+    }
+
+    @Override
+    public RecordsWithSplitIds<Elasticsearch7Record<T>> fetch() throws IOException {
```

Comment:
I think you want to return the `SearchHit`s from this method to allow better scalability. The idea behind the split reader is to do the basic IO task (reading from elasticsearch) and return the elements as fast as possible so that the source readers can immediately consume them. Now the deserialization and the IO task is coupled and may lead to bad performance.

The fetch method is used internally to push the elements to a concurrent queue that is used to hand over to the split reader threads.

Reply:
Thank you for the clarification, I'll separate it :)